tests/test_vol.py

```diff
@@ -1,58 +1,706 @@
+    assert(g.levels == exp_levels)
+
+
+def test_read_levels_nofile():
+    """Read levels from nonexisting file"""
+    levelfile = ""
+    g = vol.IsoVolDatabase()
+    with pytest.raises(RuntimeError) as error_info:
+        g.read_levels(levelfile)
+    assert 'does not exist' in str(error_info)
+
+
+# Generate Levels parametrized tests:
+# linear: (6, 5, 15, 'lin', [5., 7., 9., 11., 13., 15.]
+# log: (6, 1, 1e5, 'log', [1, 10, 1e2, 1e3, 1e4, 1e5])
+# ratio, max included: (5, 1, 625, 'ratio', [1., 5., 25., 125., 625.])
+# ratio, max not included: (5, 1, 700, 'ratio', [1., 5., 25., 125., 625.])
+@pytest.mark.parametrize("N,minN,maxN,mode,exp",
+                         [(6, 5, 15, 'lin', [5., 7., 9., 11., 13., 15.]),
+                          (6, 1, 1e5, 'log', [1, 10, 1e2, 1e3, 1e4, 1e5]),
+                          (5, 1, 625, 'ratio', [1., 5., 25., 125., 625.]),
+                          (5, 1, 700, 'ratio', [1., 5., 25., 125., 625.])])
+def test_generate_levels(N, minN, maxN, mode, exp):
+    """generate levels with different modes"""
+    g = vol.IsoVolDatabase()
+    g.generate_levels(N, minN, maxN, mode=mode)
+    assert(g.levels == exp)
+
+
+def test_generate_levels_error():
+    """generate levels with invalid mode"""
+    g = vol.IsoVolDatabase()
+    with pytest.raises(RuntimeError) as error_info:
+        g.generate_levels(6, 5, 1e5, mode='nonsense')
+    assert 'Level generation' in str(error_info)
+
+
+# Generate Volumes parametrized tests:
+#   (1) Min and Max w/in data bounds
+#   (2) Mid level no data
+#   (3) Min and max out of data bounds
+@pytest.mark.parametrize("levels,id", [([5, 15, 25, 35], 1),
+                                       ([5, 15, 25, 28, 35], 2),
+                                       ([-5, 5, 15, 25, 35, 45], 3)])
+def test_generate_volumes(levels, id):
+    """Generate all isovolume files with different bounds.
+    Some should raise warnings. Make sure generated files are correct."""
+    # Generate the volumes
+    g = vol.IsoVolDatabase()
+    db = test_dir + "/test-gen-vols-{}".format(id)
+    if os.path.isdir(db):
+        shutil.rmtree(db)
+    # data out of range should produce warning
+    with pytest.warns(None) as warn_info:
+        g.generate_volumes(test_mesh, dataname, dbname=db, levels=levels)
+    # assert number of warnings raised
+    if id == 2:
+        # no data found warning
+        assert(len(warn_info) == 1)
+    elif id == 3:
+        # levels are out of bounds
+        assert(len(warn_info) == 2)
+    else:
+        assert(len(warn_info) == 0)
+    # check that files produced are the same
+    gen_vols_dir = db + "/vols"
+    levelfile_out = db + "/levelfile"
+    res = filecmp.cmpfiles(exp_vols_dir, gen_vols_dir, common_files)
+    match_list = res[0]
+    non_match = res[1]
+    assert(match_list == common_files)
+    assert(non_match == [])
+    # check that the level files are the same
+    res = __compare_levelfiles(exp_levelfile, levelfile_out)
+    assert(res)
```

Comment:
Instead of having multiple `assert` lines distributed within a test,
set a flag for each `assert` line and at the end, using those flags, trigger a single `assert`.